src/main/java/io/inscopemetrics/kairosdb/HistogramDataPointV2Factory.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2019 Inscope Metrics, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.inscopemetrics.kairosdb;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import io.inscopemetrics.kairosdb.proto.v2.FormatV2;
+import org.kairosdb.core.DataPoint;
+import org.kairosdb.core.datapoints.DataPointFactory;
+
+import java.io.DataInput;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+
+/**
+ * Factory that creates {@link HistogramDataPointV2Impl}.
+ *
+ * @author Brandon Arp (brandon dot arp at inscopemetrics dot io)
+ */
+public class HistogramDataPointV2Factory implements DataPointFactory {
+    /**
+     * Name of the Data Store Type.
+     */
+    public static final String DST = "kairos_histogram_v2";
+    /**
+     * Name of the group type.
+     */
+    public static final String GROUP_TYPE = "histogram";
+
+    /**
+     * Default constructor.
+     */
+    public HistogramDataPointV2Factory() { }
+
+    @Override
+    public String getDataStoreType() {
+        return DST;
+    }
+
+    @Override
+    public String getGroupType() {
+        return GROUP_TYPE;
+    }
+
+    @Override
+    public DataPoint getDataPoint(final long timestamp, final JsonElement json) {
+        final TreeMap<Double, Long> binValues = new TreeMap<>();
+
+        final JsonObject object = json.getAsJsonObject();
+        final double min = object.get("min").getAsDouble();
+        final double max = object.get("max").getAsDouble();
+        final double mean = object.get("mean").getAsDouble();
+        final double sum = object.get("sum").getAsDouble();
+        final JsonObject bins = object.get("bins").getAsJsonObject();
+
+        for (final Map.Entry<String, JsonElement> entry : bins.entrySet()) {
+            binValues.put(Double.parseDouble(entry.getKey()), entry.getValue().getAsLong());
+        }
+
+        final byte precision = Optional.ofNullable(object.get("precision")).map(JsonElement::getAsByte).orElse((byte) 7);
+        return new HistogramDataPointV2Impl(timestamp, precision, binValues, min, max, mean, sum);
+    }
+
+    @Override
+    public DataPoint getDataPoint(final long timestamp, final DataInput buffer) throws IOException {
+        final int length = buffer.readInt();
+        final byte[] data = new byte[length];
```

Comment:
Looks like the `int` primitive is a signed int32, so this allocation can be up to 2GiByte. Should there be a smaller limit on the size of the serialized form of a DataPoint (which you would, presumably, enforce here)?

Reply:
Are you concerned about possible corruption and thus over-reads of the buffer? Or is this more performance related, to limit the maximum size of histogram written.

The latter case I believe should really be handled on write, although I don't know what a reasonable value would be.  The former is trickier. I guess ideally, we should write some hash or signature to the stream if we are worried about corruption. However, (and I have not dug into what actually exists) I would hope that the underlying data store (of which this is agnostic) already provides some data consistency and anti-corruption features.

Let me know if you are asking for some other reason.

Reply:
> Or is this more performance related, to limit the maximum size of histogram written.

This one! When possible, I try to make the memory usage of code predictable, for two reasons:
1. If histograms are of bounded size, then you can straightforwardly reason from "number of histograms stored in memory at once" to derive the memory (RSS limit, JVM heap limit, whatever) required for the service to run without crashing.
2. As deployed, our system almost certainly makes some implicit assumptions (via the existing process-wide memory limits) about how big a histogram can be, both on the wire and in memory - probably hundreds of KiBytes at most. If someone ever starts sending multi-gigabyte histograms, or simply sends a lot of requests with size headers that result in a huge buffer here (bitflip? buggy sender? malicious sender that we hooked up directly to KairosDB for some reason?), then suddenly those assumptions will get violated, resulting in crash. I think it's best to avoid creating such opportunities for _surprisingly, radically larger_ resource consumption.

Reply:
Histograms are not bounded by size. I agree, this would be a nice characteristic to have; however, it was traded explicitly against predictable error (vs say HDR Histogram). As long as this fundamental property is in place, I don't think it makes sense to put limits on what can be read/written in KDB. 

The system definitely makes assumptions about scale based on its configuration (e.g. allocated heap and concurrency) and deployment (e.g. number of procs and hosts). These are the knobs we have to turn to scale the system and adjust its runtime characteristics to load. Over time, I expect we will find the most efficient use of a single host for each component and then scale "linearly" simply by adding more hardware.

The system is definitely open to attack and client-side bugs causing issues. In my opinion, these should be defended against at the border. What's the point in restricting histogram size on read after it's already tanked the rest of the system? There is an argument for defense in-depth, but we should probably start with the most likely vectors first.

If/When we do add such restrictions it should probably be on the write side and not on the read side of KDB. Perhaps even in the API layer. It needs to be informed by something tangible, otherwise we will simply drop valid data and increase the threshold. Ultimately, the threshold still doesn't provide the fixed-space guarantee that you seek; histograms are read and aggregated over unbound time intervals by users in KDB resulting in larger and larger histograms which can blow-up the query part of the stack.

Again, not saying we shouldn't do this, but it's not as simple as failing to read if there are too many buckets.